### PR TITLE
Make memory_consumption more generic

### PR DESCRIPTION
**Pull request by drwells**
`memory_consumption.cc` currently contains one function defined for compatibility with an older version of IBM's `xlc`. I removed it and used `type_traits` to get rid of a bunch of boiler plate while I was at it.

It would be nice to have one overload of `memory_consumption` for all fundamental types, but I do not believe that is possible (functions can't be partially specialized).

### CODE REVIEW

File: include/deal.II/base/memory_consumption.h [hunk 1]
```diff
@@ -505,12 +471,20 @@ namespace MemoryConsumption
...
-    return mem;
+    // shortcut for types that do not allocate memory themselves
+    if (std_cxx11::is_fundamental<T>::value || std_cxx11::is_pointer<T>::value)
+      {
+        return v.capacity()*sizeof(T) + sizeof(v);
+      }
+    else
+      {
+        std::size_t mem = sizeof(std::vector<T>);
+        const unsigned int n = static_cast<unsigned int>(v.size());
```
Comment (tjhei) [inline include/deal.II/base/memory_consumption.h]: Can we remove the cast from size_t to unsigned int here?

**Comment by drwells** [comment]
Of course I figure out how to get the generic version working right after I posted this. I will fix that and the cast in the morning.

**Comment by drwells** [comment]
I see no reason why we need that `static_cast` in there (they are both unsigned types) so yes, lets get rid of it.

I got rid of a lot of boiler plate with a little bit of `enable_if`.

**Comment by tjhei** [comment]
./ did you check that this works without cxx11 support?
./ can we move enable_if inside base/std_cxx11/ and do a "using boost::enable_if_c" if we have no c++11?

**Comment by tjhei** [comment]
./ we don't have any test for memory_consumption for basic types, are you willing to add some especially for what is new?
./ do the tests base/memory_consumption_xx pass?

**Comment by drwells** [comment]
@tjhei, in order:
1. Yes, all of the tests work with and without C++11.
2. I do not believe this is possible. The problem with that is that the C++11 `enable_if` is equivalent to `boost::enable_if_c`, *not* `boost::enable_if`, and one cannot alias templates without C++11 support.
3. I added a test for fundamental types. I am a little hesitant to add more tests because the memory consumption of things like `std::string` depends on the implementation.
4. Of course the tests pass :)

**Comment by tjhei** [comment]
Great, thanks. Looks good to me. Does anybody else have a comment?

**Comment by bangerth** [comment]
Less code = more better.

**Comment by bangerth** [comment]
As for the difference between std::enable_if and boost::enable_if: we're not bound to simply import things into `namespace std_cxx11`. I see nothing wrong with something of the following kind:
```
#if C++11
  using std::enable_if;
#else
  template <bool B, class T = void> struct enable_if : public boost::enable_if_c<B,T> {};
#endif
```
I think this achieves what we want, right?

**Pull request merged**